plugin/trino-ldap-groups-provider/src/test/java/io/trino/plugin/ldapgroups/TestLdapGroupsProviderIntegration.java

```diff
@@ -0,0 +1,270 @@
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.testcontainers.containers.BindMode;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@Testcontainers
+public class TestLdapGroupsProviderIntegration
+{
+    private static final int OPENLDAP_EXPOSED_PORT = 389;
+
+    private final LdapGroupsProviderFactory factory = new LdapGroupsProviderFactory();
+
+    @Container
+    private static final GenericContainer<?> openldapContainer = new GenericContainer<>("osixia/openldap:1.5.0")
+            .withNetworkAliases("openldap")
+            .withEnv("LDAP_ORGANISATION", "Trino")
+            .withEnv("LDAP_DOMAIN", "trino.io")
+            .withEnv("LDAP_TLS", "false")
+            .withEnv("LDAP_SEED_INTERNAL_LDIF_PATH", "/ldap/ldif")
+            .withExposedPorts(OPENLDAP_EXPOSED_PORT)
+            .withFileSystemBind(
+                    Resources.getResource("ldap-trino-io.ldif").getPath(),
+                    "/ldap/ldif/ldap-trino-io.ldif",
+                    BindMode.READ_ONLY)
+            // The search query is relevant since cn=qa,ou=groups,dc=trino,dc=io is the last document in the ldif file
+            .waitingFor(Wait.forSuccessfulCommand(
+                    "if [ $(ldapsearch -LLL -x -b dc=trino,dc=io -H ldap:// -D cn=admin,dc=trino,dc=io -w admin cn=qa | wc -c) -eq 0 ]; then exit 1; fi"));
+
+    private static Map<String, String> baseConfig;
+
+    @BeforeAll
+    public static void setup()
+    {
+        baseConfig = Map.of(
+                "ldap.url", "ldap://localhost:" + openldapContainer.getMappedPort(OPENLDAP_EXPOSED_PORT),
+                "ldap.allow-insecure", "true",
+                "ldap.admin-user", "cn=admin,dc=trino,dc=io",
+                "ldap.admin-password", "admin",
+                "ldap.user-base-dn", "ou=users,dc=trino,dc=io",
+                "ldap.group-base-dn", "ou=groups,dc=trino,dc=io",
+                "ldap.cache-ttl", "0s");
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideForTestGetGroupsOnlyGroupsForAdmin")
+    public void testGetGroupsOnlyGroupsForAdmin(String userName, Set<String> expectedGroups)
+    {
+        var config = ImmutableMap.<String, String>builder()
+                .putAll(baseConfig)
+                .put("ldap.user-search-filter", "(uid={0})")
```

Comment:
Can we make the search filter similar to what we configure in the Authenticator - Like in case of LdapAuthenticator - we use `${USER}` which will replace the username pattern. Can we reuse the same approach here ?

Reply:
This is using `{0}` so that the substitution can be done by the LDAP client/server, compared to the LDAP password authentication which handles substitution on its own.

This is also the reason behind the change to [`JdkLdapClient`](https://github.com/trinodb/trino/pull/20157/files#diff-c99031194fd25f675a123d6f9648f402e0b0f4dc3356cf574c31459ed96bd607R181).

I would expect this approach to be more secure, but I can't find much documentation on it.